### DISCUSSION TITLE
Change-namespaced short commit IDs

### DISCUSSION THREAD

**Discussion by mlcui-corp** [jj-vcs/jj#4312] [Ideas]
At the moment, commit IDs (cf. change IDs) are typically used for:

- Manipulating (abandoning) divergent changes - two visible commits with the same change ID.
- Manipulating (restoring) old versions of changes, found through `jj obslog`, `jj --at-op`, or an old invocation of `jj log`.
- Git interoperability, for commands such as `git push origin COMMIT_ID:refs/for/main`. These cannot use short IDs.

The current status quo of short commit IDs - unique prefixes for all visible commits in `revsets.short-prefixes`, then globally-unique prefixes - is fine for the divergent change use case, but falls short for the "old versions of changes" use case:

- Short commit IDs from the output of an old `jj log` invocation (equivalently, `jj log --at-op`) often don't work, as those commits became invisible. Restoring those require a longer prefix.
- Short commit IDs for hidden commits - such as the ones in `jj obslog` - are very long, as they need to be globally unique prefixes.

My proposal would be to allow specification of short commit IDs in the namespace of a given change. Using a (not final) syntax of `change??commit`, `tnpw??a8` would refer to "the unique (visible) commit in the obslog of change `tnpw` with prefix `a8`".

If such a thing existed:

- Divergent changes in `jj log` could be shown as `change??short_visible_commit` instead of the current `change??`. As there are commonly only two visible commits for a divergent change, `short_visible_commit` will often only be a single character.
- Specifying a commit from a divergent change might be more intuitive - you specify the change, then which commit to refer to.
- Commit IDs could be shown as `change??short_commit` (not over visible commits), so short commit IDs from old `jj log` outputs should still work even if the commit was hidden (as long as there are no new commits with the same change and prefix).
- `jj obslog` commit IDs would be much shorter.

**Comment by hotsphink** [comment]
Can we use の (i.e., changeのcommit)? :-)

**Comment by DavidDTA** [comment]
More modern versions of `jj` now use `change/N` instead of `change??` to display divergent commits, but I think this proposal is even better since the change referred to by `change/N` can change since it's based on last modified time

**Reply by PhilipMetzger** [reply]
> but I think this proposal is even better since the change referred to by `change/N` can change since it's based on last modified time

No, they use the index position which is determined by timestamp in the Git backend.

**Reply by DavidDTA** [reply]
Thanks for the clarification. My point stands, though, that because they are indexed in reverse chronological order according to that timestamp, the same change offset can refer to a different commit after additional operations have occurred
